src/main/cpp/lang/js.cpp

```diff
@@ -1615,10 +1615,102 @@ struct [[nodiscard]] Highlighter : Highlighter_Base {
         if (number.erroneous) {
             emit_and_advance(number.length, Highlight_Type::error);
         }
-        else {
-            // TODO: more granular output
-            emit_and_advance(number.length, Highlight_Type::number);
+        const std::size_t start = index;
+        if (number.prefix > 0) {
+            emit_and_advance(number.prefix, Highlight_Type::number_decor);
+        }
+        if (number.integer > 0) {
+            std::size_t chars = 0;
+            auto flush_digits = [&] {
+                if (chars > 0) {
+                    emit(index - chars, chars, Highlight_Type::number);
+                    chars = 0;
+                }
+            };
+
+            std::size_t remaining = number.integer;
+            while (remaining > 0) {
```

Comment:
It would be a lot more obvious what happens here if you wrote it as a three-statement `for`:
```suggestion
            for (std::size_t remaining = number.integer; remaining > 0; --remaining) {
```